lib/datadog/grok/src/parse_grok_rules.rs

```diff
@@ -0,0 +1,375 @@
+        .iter()
+        .filter(|&r| !r.is_empty())
+        .map(|r| {
+            parse_grok_rule(
+                r,
+                &mut prev_rule_definitions,
+                &mut prev_rule_filters,
+                &mut grok,
+            )
+        })
+        .collect::<Result<Vec<GrokRule>, Error>>()
+}
+
+fn parse_grok_rule<'a>(
+    rule: &'a str,
+    mut prev_rule_definitions: &mut HashMap<&'a str, String>,
+    mut prev_rule_filters: &mut HashMap<&'a str, HashMap<LookupBuf, Vec<GrokFilter>>>,
+    grok: &mut Grok,
+) -> Result<GrokRule, Error> {
+    let mut split_whitespace = rule.splitn(2, ' ');
+    let split = split_whitespace.by_ref();
+    let rule_name = split.next().ok_or_else(|| {
+        Error::InvalidGrokExpression(
+            rule.to_string(),
+            "format must be: 'ruleName definition'".into(),
+        )
+    })?;
+    let mut rule_def = split
+        .next()
+        .ok_or_else(|| {
+            Error::InvalidGrokExpression(
+                rule.to_string(),
+                "format must be: 'ruleName definition'".into(),
+            )
+        })?
+        .to_string();
+
+    let rule_def_cloned = rule_def.clone();
+    lazy_static! {
+        static ref GROK_PATTER_RE: Regex = Regex::new(r"%\{.+?\}").unwrap();
+    }
+    // find all patterns %{}
+    let raw_grok_patterns = GROK_PATTER_RE
+        .find_iter(rule_def_cloned.as_str())
+        .map(|rule| rule.as_str())
+        .collect::<Vec<&str>>();
+    // parse them
+    let mut grok_patterns = raw_grok_patterns
+        .iter()
+        .map(|pattern| {
+            parse_grok_pattern(pattern)
+                .map_err(|e| Error::InvalidGrokExpression(pattern.to_string(), e.to_string()))
+        })
+        .collect::<Result<Vec<GrokPattern>, Error>>()?;
+
+    grok_patterns = index_repeated_fields(grok_patterns);
+
+    let mut filters: HashMap<LookupBuf, Vec<GrokFilter>> = HashMap::new();
+    let pure_grok_patterns: Vec<String> = grok_patterns
+        .iter()
+        .map(|pattern| {
+            purify_grok_pattern(
+                &pattern,
+                &mut filters,
+                &mut prev_rule_filters,
+                &mut prev_rule_definitions,
+            )
+        })
+        .collect::<Result<Vec<String>, Error>>()?;
+
+    // replace grok patterns with "purified" ones
+    let mut pure_pattern_it = pure_grok_patterns.iter();
+    for r in raw_grok_patterns {
+        rule_def = rule_def.replace(r, pure_pattern_it.next().unwrap().as_str());
+    }
+
+    // collect all filters to apply later
+    grok_patterns
+        .iter()
+        .filter(|&rule| {
+            rule.destination.is_some() && rule.destination.as_ref().unwrap().filter_fn.is_some()
+        })
+        .map(|rule| {
+            let dest = rule.destination.as_ref().unwrap();
+            let filter = GrokFilter::try_from(dest.filter_fn.as_ref().unwrap())?;
+            Ok((dest, filter))
+        })
+        .collect::<Result<Vec<_>, Error>>()?
```

Comment:
But given all that, I think this whole chain could be reduced to a single loop which means it could avoid having to create this intermediate `Vec`.

Reply:
Indeed! Simplified in https://github.com/timberio/vector/pull/8850/commits/9cd2497b46916033b852b67a45e7b6c1f9c66916.